### ISSUE TITLE
fd_read: is this `+` supposed to be there?

### ISSUE THREAD

**Issue by corwin-of-amber** [wasmerio/wasmer-js#135]
This line: 
https://github.com/wasmerio/wasmer-js/blob/b55bab0130bf7c35bcc7fe6d319202dbc02e5200/packages/wasi/lib/wasi.ts#L725

Occurs in the import `fd_read`. Since it computes a length, does it even make senes to add the offset?

Removing ` + stats.offset` worked better for me.

**Comment by torch2424** [comment]
@syrusakbary Would you have more context on this? Otherwise I can do the investigation 😄

**Comment by syrusakbary** [comment]
Hi @corwin-of-amber,

Thanks for creating the issue. We recently added support for `fd_seek` ( #131 ) and it required this changes for `fd_read` to work properly.

In the case that is not working for you... can you post here the WASM where it breaks so we can take a look?

PS: we also need to add more testing to test WASI wasm files to make sure everything is working as it should, if you want to help us on that, that would be awesome!

**Comment by corwin-of-amber** [comment]
I will, it will take me some time, basically I was compiling this small routine:
```c
#include <stdio.h>
#include <string.h>
#include <unistd.h>

char buffer[2048];

void cmdloop() {
    setvbuf(stdin, 0, _IONBF, 0);  // unbuffered

    for (int i = 0; i < 2; i++) {
        fgets(buffer, sizeof(buffer), stdin);
        
        printf("buffer=%s\n", buffer);
    }
}
```

And running it with the example code at https://github.com/wasmerio/wasmer-js/blob/master/examples/node-stubbed/index.js. I added debug prints in stdinRead and noticed that the length is sometimes larger than the size of the stdinBuf.

**Comment by corwin-of-amber** [comment]
Attaching C, JS, and WASM. The C program was compiled with:
```/opt/wasi-sdk/bin/clang busy.c -o busy-wasi.wasm```

[busy.zip](https://github.com/wasmerio/wasmer-js/files/3773766/busy.zip)

**Comment by TJKoury** [comment]
This is related to my issue #145, removing the offset fixes the issue.

**Comment by syrusakbary** [comment]
Thanks for the report. We are working on a fix now :)

**Comment by torch2424** [comment]
@syrusakbary #146 fixes this correct? 😄 Ccan release a patch for @TJKoury and @corwin-of-amber ? 😄

**Comment by syrusakbary** [comment]
@torch2424 that's right... this should be fixed by #146.

Let's release the next version soon (after we add the CLI): 0.4.0

**Comment by torch2424** [comment]
@syrusakbary Sounds good! Also, it may be good to take a look at #146 , since @corwin-of-amber made a comment we may want to address before releasing 😄

**Comment by syrusakbary** [comment]
Cool, this should be now fixed in `0.4.0`. Closing the issue.
If you have further issues with this, please let us know and reopen the issue! :)
